fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java

```diff
@@ -537,12 +556,88 @@ public static Map<String, Range<PartitionKey>> getRangePartitionMapOfExternalTab
                     isConvertToDate ? PrimitiveType.DATE : partitionColumn.getPrimitiveType());
 
-            Preconditions.checkState(!mvPartitionRangeMap.containsKey(lastPartitionName));
-            mvPartitionRangeMap.put(lastPartitionName, Range.closedOpen(lastPartitionKey, endKey));
+            putRangeToMvPartitionRangeMap(mvPartitionRangeMap, lastPartitionName, lastPartitionKey, endKey);
+        }
+        return mvPartitionRangeMap;
+    }
+
+    /**
+     * Map partition values to partition ranges, eg:
+     * [1992-01-01,1992-01-02,1992-01-03]
+     * ||
+     * \/
+     * [0000-01-01, 1992-01-02),[1992-01-01, 1992-01-02),
+     * [1992-01-02, 1992-01-03),[1993-01-03, 1993-01-04),
+     * [1993-01-03, 9999-12-31)
+     * <p>
+     * NOTE:
+     * External tables may contain multi partition columns, so the same mv partition name may contain multi external
+     * table partitions. eg:
+     * partitionName1 : par_col=0/par_date=2020-01-01 => p20200101
+     * partitionName2 : par_col=1/par_date=2020-01-01 => p20200101
+     */
+    public static Map<String, Range<PartitionKey>> getRangePartitionMapOfJDBCTable(Table table,
+                                                                                   Column partitionColumn,
+                                                                                   List<String> partitionNames,
+                                                                                   Expr partitionExpr)
+            throws AnalysisException {
+
+        // Determine if it is the str2date function
+        boolean isConvertToDate = isConvertToDate(partitionExpr, partitionColumn);
+        List<PartitionKey> partitionKeys = new ArrayList<>();
+        Map<String, PartitionKey> mvPartitionKeyMap = Maps.newHashMap();
+        for (String partitionName : partitionNames) {
+            partitionName = transformMaxValueForJDBCTable(partitionColumn, partitionName);
+            putMvPartitionKeyIntoMap(table, partitionColumn, partitionKeys, mvPartitionKeyMap, partitionName);
         }
+
+        LinkedHashMap<String, PartitionKey> sortedPartitionLinkMap = mvPartitionKeyMap.entrySet().stream()
+                .sorted(Map.Entry.comparingByValue(PartitionKey::compareTo))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
+        int index = 0;
+        PartitionKey lastPartitionKey = null;
+        String partitionName = null;
+
+        Map<String, Range<PartitionKey>> mvPartitionRangeMap = new LinkedHashMap<>();
+        for (Map.Entry<String, PartitionKey> entry : sortedPartitionLinkMap.entrySet()) {
+            // Adapt to the range partitioning method of JDBC Table, the partitionName adopts the name of upperBound
+            partitionName = entry.getKey();
+            // If the partition column type is int, it needs to start from 0, and if type is data, it needs to start from 0000-01-01
+            if (index == 0) {
```

Comment:
use lastPartitionKey == null directly and remove index?

Reply:
fixed it

Reply:
This modification is more beautiful